### PR TITLE
[+breaking] Several architectural tweaks to the scheduler

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
After some mental experiments I came to conclusion that
https://github.com/sociomantic-tsunami/ocean/issues/341 is not directly a
problem, and what was observed as such, was in fact result of memory corruption
caused by https://github.com/sociomantic-tsunami/ocean/issues/353

However, I also came to conclusion that there are possible other issues if it is
allowed to resume tasks from termination hooks, because it will leave "finished"
tasks running for prolonged amount of time. So I still used original idea to fix
that.

Lot of other changes came as a side effect from digging through the scheduler
sources, most notably introduction of `IScheduler` interface. Fortunately, most
of those don't make any impact on user-facing API so expected breakage impact is
very low. Still, it requires marking next minor release as `+breaking` per Neptune.

### CODE REVIEW

File: relnotes/isheduler.feature.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+* `ocean.task.IScheduler`
+
+  New module defining all basic data types needed to interact with the scheduler
+  API. It is recommended that only module that needs to configure the scheduler
+  would import original `ocean.task.Scheduler` and all others should switch to
+  new light-weight one.
+
+  `ocean.task.IScheduler.theScheduler` will return the very same global
+  scheduler reference but downcast to `IScheduler` interface.
```
Comment (gavin-norman-sociomantic) [inline relnotes/isheduler.feature.md]: I think it's an upcast actually.

File: relnotes/isheduler.feature.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+* `ocean.task.IScheduler`
```
Comment (gavin-norman-sociomantic) [inline relnotes/isheduler.feature.md]: Commit message:
"Provides new IScheduler module which allows to make use of scheduler
facilities without directly important all of involved implementation."

"important" -> "importing"?

File: relnotes/isheduler.feature.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+  New module defining all basic data types needed to interact with the scheduler
+  API. It is recommended that only module that needs to configure the scheduler
```
Comment (gavin-norman-sociomantic) [inline relnotes/isheduler.feature.md]: "modules that need to"

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- fix 2.071 build

File: src/ocean/task/IScheduler.d [hunk 4]
```diff
@@ -241,6 +241,19 @@ public interface IScheduler
...
+        Orders scheduler to resume given task unconditionally after current
+        epoll cycle. Should be used instead of plain `Task.resume` from
+        termination hooks of other tasks.
+
+        Params:
+            task = task object to resume on next cycle
+
+    ***************************************************************************/
+
+    public void resume ( Task task );
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/IScheduler.d]: I feel this method should have a different name. `resume` is so ingrained to mean "switch execution immediately to the thing being resumed" that I think it's deceptive to use the word in a different way here.

Nothing really catchy comes to mind immediately, though. `resumeAfterProcessingEvents`? :/

File: src/ocean/task/IScheduler.d [hunk 5]
```diff
@@ -241,6 +241,19 @@ public interface IScheduler
 
     /***************************************************************************
 
+        Orders scheduler to resume given task unconditionally after current
+        epoll cycle. Should be used instead of plain `Task.resume` from
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/IScheduler.d]: Is there any way to enforce this?

File: src/ocean/task/Scheduler.d [hunk 6]
```diff
@@ -409,9 +409,12 @@ final class Scheduler : IScheduler
         assert (context !is null);
         assert (context !is task);
 
-        task.terminationHook(&context.resume);
+        // this methods stack is guaranteed to still be valid by the time
+        // task finishes, so we can reference `task` from delegate
+        task.terminationHook({ theScheduler.resume(context); });
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Scheduler.d]: This commit seems more like "Implement `IScheduler.resume` method and adapt termination hooks to use it".

File: relnotes/scheduler-exceptions.migration.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# New default scheduler exception handler in tests
```
Comment (gavin-norman-sociomantic) [inline relnotes/scheduler-exceptions.migration.md]: `##`

File: relnotes/scheduler-exceptions.migration.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+Previously, default exception handler in the scheduler was simply rethrowing any
+exception in both test and actual builds. This was most backwards compatible
```
Comment (gavin-norman-sociomantic) [inline relnotes/scheduler-exceptions.migration.md]: "simply rethrowing any exception (in both test builds and actual builds)"

File: relnotes/scheduler-exceptions.migration.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+behaviour but with an unfortunate effect that unexpected exceptions in tests
+could be lost and were hard to debug in general.
+
+New default exception handler is defined to:
+
+  1) Print exception information
+  2) Call C `abort` to terminate the process immediately
```
Comment (gavin-norman-sociomantic) [inline relnotes/scheduler-exceptions.migration.md]: The way this is worded, it sounds like this is also the case in non-test builds.

File: relnotes/resume.migration.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# Disallow direct resume in termination hooks
```
Comment (gavin-norman-sociomantic) [inline relnotes/resume.migration.md]: `##`

File: src/ocean/task/IScheduler.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/task/IScheduler.d]: I see this is done in a later commit.

File: relnotes/resume.migration.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+
+It is now necessary to use `IScheduler.resume` method if one task has to be
```
Comment (gavin-norman-sociomantic) [inline relnotes/resume.migration.md]: This commit should also edit the method comments added previously that said "you shouldn't do this". (It's now enforced, not merely prohibited by convention.)

File: relnotes/select-fiber-ext.deprecation.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# Deprecate transitional `SelectFiberSupport` task extenstion
```
Comment (gavin-norman-sociomantic) [inline relnotes/select-fiber-ext.deprecation.md]: `##` :D

File: relnotes/select-fiber-ext.deprecation.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+
+This extension was intended as a temporary hack to simplify initial switching to
+task system and is known to be rather unreliable/buggy. Currently task system is
+supported wide enough to make this hack unnecessary anymore.
```
Comment (gavin-norman-sociomantic) [inline relnotes/select-fiber-ext.deprecation.md]: "widely"

File: relnotes/await-timeout.feature.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Call `awaitOrTimeout` via scheduler interface
+
+Thanks to introducing `IScheduler` interface, it is not possible to call
```
Comment (gavin-norman-sociomantic) [inline relnotes/await-timeout.feature.md]: "it is now possible"

File: relnotes/await-timeout.feature.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
+# Call `awaitOrTimeout` via scheduler interface
```
Comment (gavin-norman-sociomantic) [inline relnotes/await-timeout.feature.md]: `##`

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Looks good generally.

Apart from the `##` thing, I think your release notes are missing the `name.of.affected.module` part.

File: src/ocean/task/IScheduler.d [hunk 16]
```diff
@@ -0,0 +1,325 @@
+/*******************************************************************************
+
+    This module defines public API for `ocean.task.Scheduler` together with
+    all involved data types. It is intended to be used in other modules instead
+    of direct `ocean.task.Scheduler` import to untangle complex dependency chain
+    the latter brings.
+
+    Copyright:
+        Copyright (c) 2017 sociomantic Labs gmbH. All rights reserved.
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/IScheduler.d]: you got capitalization wrong ;-) `sociomantic labs GmbH`

File: src/ocean/task/IScheduler.d [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+
+import core.thread;
+
+import ocean.task.Task;
+import ocean.meta.traits.Indirections : hasIndirections;
+import ocean.io.select.EpollSelectDispatcher;
+
+/*******************************************************************************
+
+    Public methods of `ocean.task.Scheduler` stripped from implementation
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/IScheduler.d]: It feels weird to mention the concrete implementation of the interface as the docs for the interface. Maybe you move the scheduler's docs here and then adjust scheduler's docs to be "implementation of IScheduler with ..."

File: src/ocean/task/IScheduler.d [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+
+        /// Defines single mapping of `ClassInfo` to worker fiber pool in
+        /// configuration
+        public struct PoolDescription
+        {
+            /// result of `Task.classinfo` for task type which is to be handled
+            /// by this pool
+            ClassInfo task_kind;
+
+            /// worker fiber allocate stack size
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/IScheduler.d]: `worker fiber's allocated stack size.`

File: src/ocean/task/IScheduler.d [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+        size_t suspended_queue_busy;
+        size_t suspended_queue_total;
+        size_t worker_fiber_busy;
+        size_t worker_fiber_total;
+    }
+
+    /***************************************************************************
+
+        Getter for scheduler epoll instance. Necessary for integration with
+        `ISelectClient` utilities so that new select event can be registered.
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/IScheduler.d]: `select event` or `select client`?

File: src/ocean/task/internal/FiberPoolWithQueue.d [hunk 20]
```diff
@@ -68,7 +69,7 @@ public class FiberPoolWithQueue : FiberPool
     **************************************************************************/
 
     public this ( size_t queue_limit, size_t stack_size, size_t limit,
-        void delegate(istring, int) process_events )
+        void delegate () process_events )
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/internal/FiberPoolWithQueue.d]: These changes don't look connected with this commit?

File: src/ocean/task/IScheduler.d (hunk 4 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/task/IScheduler.d]: `delayedResume`? `resumeDelayed`?

File: relnotes/scheduler-exceptions.migration.md (hunk 9 above)
Comment (nemanja-boric-sociomantic) [reply relnotes/scheduler-exceptions.migration.md]: Yes, it should mention `version (UnitTest)` (also commit message: "Change the default scheduler exception handler in unittest buids").

File: src/ocean/task/Scheduler.d [hunk 21]
```diff
@@ -861,6 +868,28 @@ public void initScheduler ( SchedulerConfiguration config,
...
+                    "Unhandled exception in task {} ({})",
+                    cast(void*) t, t.classinfo.name
+                );
+            }
+            else
+            {
+                Stdout.formatln("Unhandled exception in epoll/scheduler");
+            }
+
+            Stdout.formatln("\t{} ({}:{})", e.message(), e.file, e.line)
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/Scheduler.d]: Maybe `Stderr` is better choice?

File: src/ocean/io/select/protocol/task/TaskSelectTransceiver_test.d [hunk 22]
```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/io/select/protocol/task/TaskSelectTransceiver_test.d]: bad copyright capitalisation, missing module header description if you want one.

File: src/ocean/task/IScheduler.d [hunk 23]
```diff
@@ -202,19 +202,6 @@ public interface IScheduler
 
     /***************************************************************************
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/task/IScheduler.d]: This commit can be easily squashed with the first IIUC.

File: Build.mak [hunk 24]
```diff
@@ -59,25 +59,11 @@ TEST_FILTER_OUT += \
...
-$O/test-filesystemevent: override LDFLAGS += -lrt
-
-$O/test-selectlistener: override LDFLAGS += -lebtree
-
-$O/test-unixlistener: override LDFLAGS += -lebtree
-
-$O/test-loggerstats: override LDFLAGS += -lebtree
-
-$O/test-signalext: override LDFLAGS += -lebtree
+$O/test-%: override LDFLAGS += -lebtree
```
Comment (nemanja-boric-sociomantic) [inline Build.mak:62]: This change deserves separate commit :+1:.

File: relnotes/await-timeout.feature.md [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+```D
+import ocean.task.IScheduler;
+theScheduler.awaitOrTimeout(task, timeout);
+```
+
+Internally it still does the same call to global function in
+`ocean.task.util.Timer`, but new alias allows for slightly more uniform API.
+
+NB: this means that any app using the scheduler is now required to add
+`-L-lebtree` to its linker flags.
```
Comment (nemanja-boric-sociomantic) [inline relnotes/await-timeout.feature.md:15]: Hm, so this is migration/feature note. Maybe add `migration` note with just this last paragraph as well.

File: src/ocean/task/IScheduler.d (hunk 4 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/task/IScheduler.d]: `delayedResume`, I am choosing you

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- rebased
- addressed all currently present comments

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @gavin-norman-sociomantic

File: relnotes/resume.feature.md [hunk 26]
```diff
@@ -0,0 +1,13 @@
+## Delayed resuming of a task
```
Comment (gavin-norman-sociomantic) [inline relnotes/resume.feature.md]: Suggest: "New method to resume a task after the event loop cycle".

File: relnotes/resume.feature.md [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+
+`ocean.task.IScheduler`
+
+`IScheduler` interface defines new method, `delayedResume`, which takes one task
+and adds it to the same resume queue as used by `processEvents` method. That
+means that such task will be resumed by the scheduler after current epoll cycle
+instead of immediate resuming by the current context.
```
Comment (gavin-norman-sociomantic) [inline relnotes/resume.feature.md]: Is it worth explicitly mentioning that the task will be resumed _from_ the epoll context, too? (Also in method comments.)

File: src/ocean/task/Scheduler.d [hunk 28]
```diff
@@ -544,6 +547,29 @@ final class Scheduler : IScheduler
...
 
+    /***************************************************************************
+
+        Orders scheduler to resume given task unconditionally after current
+        epoll cycle. Should be used instead of plain `Task.resume` from
+        termination hooks of other tasks.
+
+        Params:
+            task = task object to resume on next cycle
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/task/Scheduler.d:579]: `Throws:`

File: relnotes/scheduler-exceptions.migration.md (hunk 9 above)
Comment (gavin-norman-sociomantic) [reply relnotes/scheduler-exceptions.migration.md]: >(also commit message: "Change the default scheduler exception handler in unittest buids").

Ping.

File: relnotes/scheduler-exceptions.migration.md [hunk 29]
```diff
@@ -0,0 +1,21 @@
+## New default scheduler exception handler in tests
```
Comment (gavin-norman-sociomantic) [inline relnotes/scheduler-exceptions.migration.md:1]: This breaking change is slipped in because you're already making an important breaking change a few commits later, right?

Should the policy rather be to absolutely minimise breaking changes in minor branches? (Rather than a kind of free-for-all attitude, once one breaking change has been included ;)

File: relnotes/resume.migration.md [hunk 30]
```diff
@@ -0,0 +1,8 @@
+## Disallow direct resume in termination hooks
+
+It is now necessary to use `IScheduler.delayedResume` method if one task has to
+be resumed as a result of termination of another task. Attempting direct resume
+will result in a `SanityException` being thrown.
+
+All ocean facilities have been already adjusted to switch accordingly (for
+example, `await`).
```
Comment (gavin-norman-sociomantic) [inline relnotes/resume.migration.md]: Suggest ", but please pay attention to adjusting application code that does the same."

File: relnotes/await-timeout.migration.md [hunk 31]
```diff
@@ -0,0 +1,4 @@
+## Using `Scheduler` requires linking ebtree
```
Comment (gavin-norman-sociomantic) [inline relnotes/await-timeout.migration.md:1]: This is another breaking change in the "not essential" category.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
A few small comments.

My only significant question is the policy on including non-essential breaking changes in this PR.

File: relnotes/await-timeout.migration.md (hunk 31 above)
Comment (mihails-strasuns-sociomantic) [reply relnotes/await-timeout.migration.md:1]: This one indeed can be moved to v4.x.x (it is only needed to provide `awaitOrTimeout` via generic scheduler interface) but I took the opportunity to put it here because in practice it is almost not going to matter - pretty much any non-trivial app that uses scheduler also uses its timer utils and thus links ebtree.

File: relnotes/scheduler-exceptions.migration.md (hunk 29 above)
Comment (mihails-strasuns-sociomantic) [reply relnotes/scheduler-exceptions.migration.md:1]: No, this one is actually quite related. With old default handler migrating main breaking change is much harder because exception from failed `verify` in `Task.resume` could be lost inside epoll or replaced by some other exception during the unwinding, making it much harder to debug any test failure caused by it.

Thus this change is to actually make another breaking change less annoying.

File: relnotes/resume.feature.md (hunk 27 above)
Comment (mihails-strasuns-sociomantic) [reply relnotes/resume.feature.md]: Isn't it already implied by "same resume queue as used by `processEvents`" and "resumed by the scheduler"?

**Comment by mihails-strasuns-sociomantic** [comment]
Updated.

File: relnotes/scheduler-exceptions.migration.md (hunk 29 above)
Comment (gavin-norman-sociomantic) [reply relnotes/scheduler-exceptions.migration.md:1]: Ok :+1:

File: relnotes/resume.feature.md (hunk 27 above)
Comment (gavin-norman-sociomantic) [reply relnotes/resume.feature.md]: Ah, "resumed by the scheduler", yes.

I'd remove the bit about `processEvents`, tbh. That's an implementation detail and irrelevant to the average user. That's what threw me off.

Comment (mihails-strasuns-sociomantic) [reply relnotes/resume.feature.md]: Updated

**Review by gavin-norman-sociomantic — APPROVED** [review]
Do you have any thoughts on documenting (in the release notes) the reason why there are breaking changes in a minor release? (And the fact that they're of low impact and probably won't actually break anything.)

**Comment by mihails-strasuns-sociomantic** [comment]
I am using only available exception case - https://github.com/sociomantic-tsunami/neptune/blob/master/doc/library-user.rst#exceptions : 

> If the addressed bug is critical (i.e. may result in memory corruption, wrong business logic, etc.), it will be released as a patch release. The release notes will clearly indicate that a breaking change has been made and the tag of the patch release will have +breaking appended.

**Comment by mihails-strasuns-sociomantic** [comment]
(it only talks about patch releases but same principle applies to minor ones)

**Comment by gavin-norman-sociomantic** [comment]
Right, I was just thinking that the automatically compiled release notes will look odd to users. "Why does this contain a Migration Instructions section!?" It'll need some lead-in / explanation.

**Comment by mihails-strasuns-sociomantic** [comment]
Like a special first paragraph in mail notes?

**Comment by gavin-norman-sociomantic** [comment]
Right, and in the gitub release, I'd suggest.

Interesting... this could actually be integrated into neptune-release, based on the presence of the `+breaking` in the tag. (For the future.)

**Comment by mihails-strasuns-sociomantic** [comment]
Actually, the `+breaking` bit is currently not supported by `neptune-release` at all - I will have to do it manually this time :( (the whole release process)

**Comment by gavin-norman-sociomantic** [comment]
Ah. That's too bad. See https://github.com/sociomantic-tsunami/neptune/issues/129.

**Comment by mihails-strasuns-sociomantic** [comment]
So, with that neptune discussion in mind, is there anything pending for this PR?

**Comment by gavin-norman-sociomantic** [comment]
Given that the whole release will need to be done manually, no.
